pinot-server/src/main/java/com/linkedin/pinot/server/starter/helix/HelixServerStarter.java

```diff
@@ -164,9 +167,32 @@ public HelixServerStarter(String helixClusterName, String zkServer, Configuratio
     serverMetrics.addCallbackGauge("memory.allocationFailureCount", PinotDataBuffer::getAllocationFailureCount);
   }
 
-  private void updateInstanceConfigInHelix(int adminApiPort, boolean shuttingDown) {
+  private void waitForAllSegmentsLoaded() {
+    if (_helixServerConfig.getBoolean(CommonConstants.Server.CONFIG_OF_STARTER_ENABLE_SEGMENTS_LOADING_CHECK, CommonConstants.Server.DEFAULT_STARTER_ENABLE_SEGMENTS_LOADING_CHECK)) {
+      long startTime = System.currentTimeMillis();
+      int serverStarterTimeout = _helixServerConfig.getInt(CommonConstants.Server.CONFIG_OF_STARTER_TIMEOUT_IN_SECONDS, CommonConstants.Server.DEFAULT_STARTER_TIMEOUT_IN_SECONDS);
+      boolean allSegmentsLoaded = false;
+      while (System.currentTimeMillis() - startTime < 1000 * serverStarterTimeout) {
```

Comment:
It'd be good to firstly calculate the estimated end time beforehand, and then check `System.currentTimeMillis() < estimatedEndTime` in the while loop.

Reply:
will do